meta/tikv: Panic when s3 connection refused

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_0_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 2)
2021/08/19 17:02:40.124499 juicefs[5551] <WARNING>: upload chunks/0/107/107004_1_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_1_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 2)
2021/08/19 17:02:40.154833 juicefs[5551] <WARNING>: upload chunks/0/107/107003_3_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107003_3_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 2)
2021/08/19 17:02:40.176505 juicefs[5551] <WARNING>: upload chunks/0/107/107004_3_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_3_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 2)
2021/08/19 17:02:40.178509 juicefs[5551] <WARNING>: upload chunks/0/107/107004_2_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_2_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 2)
2021/08/19 17:02:44.293334 juicefs[5551] <WARNING>: upload chunks/0/107/107001_2_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107001_2_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.369792 juicefs[5551] <WARNING>: upload chunks/0/107/107001_3_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107001_3_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.394803 juicefs[5551] <WARNING>: upload chunks/0/107/107003_2_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107003_2_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.403039 juicefs[5551] <WARNING>: upload chunks/0/107/107003_0_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107003_0_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.415271 juicefs[5551] <WARNING>: upload chunks/0/107/107001_0_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107001_0_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.446457 juicefs[5551] <WARNING>: upload chunks/0/107/107001_1_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107001_1_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.475872 juicefs[5551] <WARNING>: upload chunks/0/107/107004_1_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_1_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.510447 juicefs[5551] <WARNING>: upload chunks/0/107/107003_1_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107003_1_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.521198 juicefs[5551] <WARNING>: upload chunks/0/107/107004_0_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_0_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.544587 juicefs[5551] <WARNING>: upload chunks/0/107/107004_2_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_2_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.558376 juicefs[5551] <WARNING>: upload chunks/0/107/107003_3_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107003_3_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:44.591641 juicefs[5551] <WARNING>: upload chunks/0/107/107004_3_16777216: RequestError: send request failed
caused by: Put "http://10.100.19.200:9000/bigtest0/jfsbigtest0/chunks/0/107/107004_3_16777216": dial tcp 10.100.19.200:9000: connect: connection refused (try 3)
2021/08/19 17:02:54.491627 juicefs[5551] <INFO>: slow operation: flush (119): OK <16.140475>
panic: 25

goroutine 906921 [running]:
github.com/juicedata/juicefs/pkg/meta.readSliceBuf(0xc0ee5e5140, 0x19, 0x20, 0xe, 0xc0ee5e5140, 0x19)
        /go/src/github.com/juicedata/juicefs/pkg/meta/slice.go:115 +0x1f6
github.com/juicedata/juicefs/pkg/meta.(*kvMeta).compactChunk(0xc000576000, 0xa, 0xaa)
        /go/src/github.com/juicedata/juicefs/pkg/meta/tkv.go:2153 +0x1f2
github.com/juicedata/juicefs/pkg/meta.(*kvMeta).compactChunk.func3(0xc000576000, 0xa, 0xaa)
        /go/src/github.com/juicedata/juicefs/pkg/meta/tkv.go:2223 +0x53
created by github.com/juicedata/juicefs/pkg/meta.(*kvMeta).compactChunk
        /go/src/github.com/juicedata/juicefs/pkg/meta/tkv.go:2220 +0x9ce


**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**

**Environment**:
- JuiceFS version (use `./juicefs --version`) or Hadoop Java SDK version:
- Cloud provider or hardware configuration running JuiceFS:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Object storage (cloud provider and region):
- Redis info (version, cloud provider managed or self maintained):
- Network connectivity (JuiceFS to Redis, JuiceFS to object storage):
- Others:

--- a/pkg/meta/redis_test.go
+++ b/pkg/meta/redis_test.go
@@ -591,6 +591,10 @@ func TestCompaction(t *testing.T) {
 	testCompaction(t, m)
 }
 
+type compactor interface {
+	compactChunk(inode Ino, indx uint32, force bool)
+}
+
 func testCompaction(t *testing.T, m Meta) {
 	_ = m.Init(Format{Name: "test"}, true)
 	var l sync.Mutex
@@ -629,13 +633,8 @@ func testCompaction(t *testing.T, m Meta) {
 	if len(cs1) != 5 {
 		t.Fatalf("expect 5 slices, but got %+v", cs1)
 	}
-	switch r := m.(type) {
-	case *redisMeta:
-		r.compactChunk(inode, 1, true)
-	case *dbMeta:
-		r.compactChunk(inode, 1, true)
-	case *kvMeta:
-		r.compactChunk(inode, 1, true)
+	if c, ok := m.(compactor); ok {
+		c.compactChunk(inode, 1, true)
 	}
 	var cs []Slice
 	_ = m.Read(ctx, inode, 1, &cs)
@@ -644,8 +643,8 @@ func testCompaction(t *testing.T, m Meta) {
 	}
 
 	// append
-	var size uint32 = 1000000
-	for i := 0; i < 50; i++ {
+	var size uint32 = 100000
+	for i := 0; i < 200; i++ {
 		var chunkid uint64
 		m.NewChunk(ctx, inode, 0, 0, &chunkid)
 		if st := m.Write(ctx, inode, 0, uint32(i)*size, Slice{Chunkid: chunkid, Size: size, Len: size}); st != 0 {
@@ -653,13 +652,8 @@ func testCompaction(t *testing.T, m Meta) {
 		}
 		time.Sleep(time.Millisecond)
 	}
-	switch r := m.(type) {
-	case *redisMeta:
-		r.compactChunk(inode, 0, true)
-	case *dbMeta:
-		r.compactChunk(inode, 0, true)
-	case *kvMeta:
-		r.compactChunk(inode, 0, true)
+	if c, ok := m.(compactor); ok {
+		c.compactChunk(inode, 0, true)
 	}
 	var chunks []Slice
 	if st := m.Read(ctx, inode, 0, &chunks); st != 0 {
@@ -672,8 +666,8 @@ func testCompaction(t *testing.T, m Meta) {
 	for _, s := range chunks {
 		total += s.Len
 	}
-	if total != size*50 {
-		t.Fatalf("size of slice should be %d, but got %d", size*50, total)
+	if total != size*200 {
+		t.Fatalf("size of slice should be %d, but got %d", size*200, total)
 	}
 
 	// TODO: check result if that's predictable

--- a/pkg/meta/sql.go
+++ b/pkg/meta/sql.go
@@ -2284,7 +2284,7 @@ func (m *dbMeta) compactChunk(inode Ino, indx uint32, force bool) {
 	if st != 0 {
 		return
 	}
-	logger.Debugf("compact %d:%d: skipped %d slices (%d bytes) %d slices (%d bytes)", inode, indx, skipped/sliceBytes, pos, len(ss), size)
+	logger.Debugf("compact %d:%d: skipped %d slices (%d bytes) %d slices (%d bytes)", inode, indx, skipped, pos, len(ss), size)
 	err = m.newMsg(CompactChunk, chunks, chunkid)
 	if err != nil {
 		logger.Warnf("compact %d %d with %d slices: %s", inode, indx, len(ss), err)

--- a/pkg/meta/sql.go
+++ b/pkg/meta/sql.go
@@ -2301,7 +2301,7 @@ func (m *dbMeta) compactChunk(inode Ino, indx uint32, force bool) {
 			return syscall.EINVAL
 		}
 
-		c2.Slices = append(append(c2.Slices[:skipped], marshalSlice(pos, chunkid, size, 0, size)...), c2.Slices[len(c.Slices):]...)
+		c2.Slices = append(append(c2.Slices[:skipped*sliceBytes], marshalSlice(pos, chunkid, size, 0, size)...), c2.Slices[len(c.Slices):]...)
 		if _, err := ses.Where("Inode = ? AND indx = ?", inode, indx).Update(c2); err != nil {
 			return err
 		}

--- a/pkg/meta/tkv.go
+++ b/pkg/meta/tkv.go
@@ -1891,7 +1891,7 @@ func (m *kvMeta) Write(ctx Context, inode Ino, indx uint32, off uint32, slice Sl
 		attr.Ctimensec = uint32(now.Nanosecond())
 		val := tx.append(m.chunkKey(inode, indx), marshalSlice(off, slice.Chunkid, slice.Size, slice.Off, slice.Len))
 		tx.set(m.inodeKey(inode), m.marshal(&attr))
-		needCompact = (len(val)/sliceBytes)%100 == 99
+		needCompact = (len(val)/sliceBytes)%50 == 49
 		return nil
 	})
 	if err == nil {
@@ -2163,23 +2163,20 @@ func (m *kvMeta) compactChunk(inode Ino, indx uint32, force bool) {
 	if st != 0 {
 		return
 	}
-	logger.Debugf("compact %d:%d: skipped %d slices (%d bytes) %d slices (%d bytes)", inode, indx, skipped/sliceBytes, pos, len(ss), size)
+	logger.Debugf("compact %d:%d: skipped %d slices (%d bytes) %d slices (%d bytes)", inode, indx, skipped, pos, len(ss), size)
 	err = m.newMsg(CompactChunk, chunks, chunkid)
 	if err != nil {
 		logger.Warnf("compact %d %d with %d slices: %s", inode, indx, len(ss), err)
 		return
 	}
 	err = m.txn(func(tx kvTxn) error {
-		buf2, err := m.get(m.chunkKey(inode, indx))
-		if err != nil {
-			return err
-		}
+		buf2 := tx.get(m.chunkKey(inode, indx))
 		if len(buf2) < len(buf) || !bytes.Equal(buf, buf2[:len(buf)]) {
 			logger.Infof("chunk %d:%d was changed %d -> %d", inode, indx, len(buf), len(buf2))
 			return syscall.EINVAL
 		}
 
-		buf2 = append(append(buf2[:skipped], marshalSlice(pos, chunkid, size, 0, size)...), buf2[len(buf):]...)
+		buf2 = append(append(buf2[:skipped*sliceBytes], marshalSlice(pos, chunkid, size, 0, size)...), buf2[len(buf):]...)
 		tx.set(m.chunkKey(inode, indx), buf2)
 		// create the key to tracking it
 		tx.set(m.sliceKey(chunkid, size), make([]byte, 8))

--- a/pkg/meta/tkv.go
+++ b/pkg/meta/tkv.go
@@ -1891,7 +1891,7 @@ func (m *kvMeta) Write(ctx Context, inode Ino, indx uint32, off uint32, slice Sl
 		attr.Ctimensec = uint32(now.Nanosecond())
 		val := tx.append(m.chunkKey(inode, indx), marshalSlice(off, slice.Chunkid, slice.Size, slice.Off, slice.Len))
 		tx.set(m.inodeKey(inode), m.marshal(&attr))
-		needCompact = (len(val)/sliceBytes)%50 == 49
+		needCompact = (len(val)/sliceBytes)%100 == 99
 		return nil
 	})
 	if err == nil {

